File: thirtyday.go
// Package main provides ...
package main

import (
	"flag"
	"fmt"
	"strings"
	"time"
)

const day = time.Hour * 24

var dated = flag.Bool("d", false, "Show start and end dates")
var shift = flag.Int("s", 0, "Number of days to shift start by")

func thirtyday(start time.Time, title string, dated bool) {
	days := []string{"M", "T", "W", "T", "F", "S", "S "}
	daycal := ""
	startstr := start.Format("2006-01-02")
	for i := 0; i < 30; i++ {
		weekday := start.Weekday()
		daycal += string(days[weekday])
		start = start.Add(day)
	}
	endstr := start.Format("2006-01-02")

	if len(title) > 0 {
		fmt.Println(title)
	}
	if dated {
		dots := strings.Repeat(".", len(daycal)-22)
		dateheader := fmt.Sprintf("%s %s %s", startstr, dots, endstr)
		fmt.Println(dateheader)
	}
	fmt.Println(daycal)
}

func main() {
	flag.Usage = func() {
		fmt.Fprintf(flag.CommandLine.Output(), "usage:\n\tthirtyday [-s <n>] [-d] <TITLE>...\n\nFlags:\n")
		flag.PrintDefaults()
	}

	flag.Parse()

	start := time.Now()
	for i := 0; i < *shift; i++ {
		start = start.Add(day)
	}

	thirtyday(start, strings.Join(flag.Args(), " "), *dated)
}

File: thirtyday.py
#!/usr/bin/env python3
from datetime import date, timedelta
from argparse import ArgumentParser
import sys


def thirtyday(*, start, title, dated):
    shift = timedelta(days=1)
    end = start + timedelta(days=30)

    days = 'MTWTFSS'
    daycal = ''
    day = start
    for _ in range(30):
        daycal += days[day.weekday()]
        if day.weekday() == 6:
            daycal += ' '
        day += shift
    dateheader = ''
    if dated:
        startstr = start.strftime("%F")
        endstr = end.strftime("%F")
        n_dots = len(daycal) - 22
        dots = "." * n_dots
        dateheader = f"{startstr} {dots} {endstr}"
    out = []
    if title:
        out.append(title)
    if dated:
        out.append(dateheader)
    out.append(daycal)
    return '\n'.join(out)


if __name__ == "__main__":
    parser = ArgumentParser(description="Get a thirtyday challenge calendar")
    parser.add_argument('-s', help='Days to shift start', type=int, default=0)
    parser.add_argument('-t', '--title',
                        help='Title to show above calendar', type=str)
    parser.add_argument('-d', '--dated',
                        help='Show start and end date', action='store_true')
    args = parser.parse_args()
    start = date.today() + timedelta(days=args.s)
    print(thirtyday(start=start, title=args.title, dated=args.dated))


